hazelcast/src/test/java/com/hazelcast/test/starter/constructor/AbstractConfigConstructor.java

```diff
@@ -105,6 +107,12 @@ static Object cloneConfig(Object thisConfigObject, ClassLoader classloader) thro
                 } else if (returnType.getName().startsWith("java") || returnType.isPrimitive()) {
                     Object thisSubConfigObject = method.invoke(thisConfigObject, null);
                     updateConfig(setter, otherConfigObject, thisSubConfigObject);
+                } else if (returnType.getName().startsWith("com.hazelcast.core.RingbufferStore")) {
+                    cloneStoreInstance(classloader, method, setter, thisConfigObject, otherConfigObject,
+                            "com.hazelcast.core.RingbufferStore");
+                } else if (returnType.getName().startsWith("com.hazelcast.core.QueueStore")) {
+                    cloneStoreInstance(classloader, method, setter, thisConfigObject, otherConfigObject,
+                            "com.hazelcast.core.QueueStore");
```

Comment:
similarly to comment above, this one will also match `com.hazelcast.core.QueueStoreFactory`

Reply:
I've added a `QueueStoreCompatibilityTest` on the EE side (with a little tweak to really test all Hazelcast versions):
```java
@RunWith(EnterpriseSerialJUnitClassRunner.class)
@Category(CompatibilityTest.class)
public class QueueStoreCompatibilityTest extends QueueStoreTest {

    @Override
    protected HazelcastInstance createHazelcastInstance(Config config) {
        // we create the factory with 3 instances, to get an Hazelcast instance with the CURRENT_VERSION as well
        // and return this to the test (instance[0] and instance[1] are proxied instances)
        TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory(3);
        HazelcastInstance[] instances = factory.newInstances(config);
        warmUpPartitions(instances);
        return instances[2];
    }
}
```
This worked well, I just made the `AbstractConfigConstructor.cloneConfig()` method more concrete about the supported types:
```java
if (returnTypeName.equals("com.hazelcast.core.RingbufferStore")
        || returnTypeName.equals("com.hazelcast.core.RingbufferStoreFactory")
        || returnTypeName.equals("com.hazelcast.core.QueueStore")
        || returnTypeName.equals("com.hazelcast.core.QueueStoreFactory")) {
    cloneStoreInstance(classloader, method, setter, thisConfigObject, otherConfigObject, returnTypeName);
}
```

There was one new error, originated in the compatibility framework: `ClassCastException: [B cannot be cast to [Ljava.lang.Object`, thrown in the `HazelcastProxyFactory.proxyObjectForStarter()` method. It was originated in the `if (arg.getClass().isArray())` code, which I pulled out in a new method. It needed a fix to support primitive arrays (I've added `byte[]`, `int[]` and `long[]` for now):
```java
private static Object toArray(ClassLoader targetClassLoader, Object arg) throws ClassNotFoundException {
    if (arg instanceof byte[]) {
        byte[] srcArray = ((byte[]) arg);
        byte[] targetArray = new byte[srcArray.length];
        arraycopy(srcArray, 0, targetArray, 0, srcArray.length);
        return targetArray;
    } else if (arg instanceof int[]) {
        int[] srcArray = ((int[]) arg);
        int[] targetArray = new int[srcArray.length];
        arraycopy(srcArray, 0, targetArray, 0, srcArray.length);
        return targetArray;
    } else if (arg instanceof long[]) {
        long[] srcArray = ((long[]) arg);
        long[] targetArray = new long[srcArray.length];
        arraycopy(srcArray, 0, targetArray, 0, srcArray.length);
        return targetArray;
    }
    Object[] srcArray = ((Object[]) arg);
    Class<?> targetClass = targetClassLoader.loadClass(srcArray.getClass().getComponentType().getName());
    Object[] targetArray = (Object[]) Array.newInstance(targetClass, srcArray.length);
    for (int i = 0; i < srcArray.length; i++) {
        targetArray[i] = proxyObjectForStarter(targetClassLoader, srcArray[i]);
    }
    return targetArray;
}
```

The `QueueStoreTest.testIssue1401QueueStoreWithTxnPoll()` fails most of the time, but it's the only one using a `TransactionalQueue`, so I guess we're still missing some kind of support for it. I've ignored it in the `QueueStoreCompatibilityTest` for now.